README.md
## flexi_image_slider

Image slider with auto scroll, dots indicator

<table>
   <tr>
      <td>
         <img width="250px" src="https://raw.githubusercontent.com/Dharini17/flexi_image_slider/master/assets/screenshot1.jpeg">
      </td>   
<td>
         <img width="250px" src="https://raw.githubusercontent.com/Dharini17/flexi_image_slider/master/assets/screenshot2.jpeg">
      </td>  
    </tr>
  <tr>
      <td>
         <img width="250px" src="https://raw.githubusercontent.com/Dharini17/flexi_image_slider/master/assets/screenshot3.jpeg">
      </td>  
    </tr>
</table>

## Features

- set image with your aspect ratio
- auto scroll option

## Installation

In your `pubspec.yaml` file within your Flutter Project:

```yaml
dependencies:
  flexi_image_slider: 1.0.5
```

## Usage

        //required fields

        aspectRatio: set you slider height like 1.0,1.5,2.0 etc...
        arrayImages: pass your List<String> type array which contains image urls
        
        //optional fields ------------
        
        autoScroll: auto scroll enable disable by set true/false here
        viewportFraction: set its value between 0.5 to 0.9 for display left & right side image preview
        boxFit: set your image content mode 
        indicatorPosition: set indicator position like none,over slider or display at bottom of slider
        indicatorAlignment: where you want to set your slider like left,right or center
        duration: set your slider image changed speed by setting duration 
        indicatorActiveColor: set active color for indicator,
        indicatorActiveColor: set deactive color for indicator,
        borderRadius: set any double value for give radius to your image
        onTap : want to execute some action on image click use this

```dart
      
        //import package
        import 'package:flexi_image_slider/flexi_image_slider.dart';


        flexi_image_slider(

            //required fields ---------
    
            context: context,
            aspectRatio: 16/14,
            arrayImages: arrayImages,
    
            //optional fields ------------
    
            autoScroll: true,
            viewportFraction: 0.8,
            boxFit: BoxFit.fill,
            indicatorPosition: IndicatorPosition.overImage,//IndicatorPosition.afterImage,IndicatorPosition.overImage,IndicatorPosition.none
            indicatorAlignment: IndicatorAlignment.left,//IndicatorAlignment.left,IndicatorAlignment.right,IndicatorAlignment.center
            duration: const Duration(seconds: 4),
            indicatorActiveColor: Colors.blue,
            indicatorDeactiveColor: Colors.grey,
            borderRadius: 0,
            onTap: (int index){
            print("$index index clicked");
            //handle your click events
            },
        ),

```

NOTE: Images used only for demo purpose

CHANGELOG.md
## 1.0.7

- indicator Active & Deactive color set

## 1.0.6

- indicator color change option
- indicator position option

## 1.0.5

- add attributes like,
- Duration
- Border radius
- Indicator
- ViewPortFraction for left,right preview
- click event


## 1.0.4

- Fixes Bugs


test/flexi_image_slider_test.dart
import 'package:flexi_image_slider/flexi_image_slider.dart';
import 'package:test/test.dart';

void main() {

}


lib/src/Utils.dart

import 'package:flutter/material.dart';
import 'package:cached_network_image/cached_network_image.dart';

class Utils{

  static CachedNetworkImage funcDisplayImage(String strImageURL, BoxFit varBoxType) {

    return CachedNetworkImage(
      imageUrl: "$strImageURL",
      imageBuilder: (context, imageProvider) => Container(
        decoration: BoxDecoration(
          image: DecorationImage(
            image: imageProvider,
            fit: varBoxType,
            // fit: BoxFit.fitWidth

            // colorFilter: ColorFilter.mode(Colors.red, BlendMode.colorBurn)
          ),
        ),
      ),
      errorWidget: (context, url, error) => Container(
        // height: 80,width: 250,
        decoration: BoxDecoration(

          shape: BoxShape.rectangle,
        ),
        child: Icon(Icons.image_rounded,color: Colors.grey,),
      ),
    );
  }

}


lib/src/Style1.dart
import 'dart:async';
import 'package:flexi_image_slider/src/Utils.dart';
import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';
import 'package:gallery_zoom_slides/gallery_zoom_slides.dart';

class Style1 extends StatelessWidget{

  final BuildContext context;
  bool autoScroll;
  final double aspectRatio;
  BoxFit boxFit;
  List<String> arrayImages;

  Style1({required this.context, required this.autoScroll, required this.aspectRatio, required this.boxFit,required this.arrayImages});

  Timer? _timer;
  final PageController _pageController = PageController(initialPage: 999,viewportFraction: 1.0);

  @override
  Widget build(BuildContext context) {

    if(autoScroll){
    _timer = Timer.periodic(Duration(seconds: 3), (Timer timer) {
      _pageController.nextPage(duration: const Duration(seconds: 1), curve: Curves.easeInOutSine);
    });}

    return

      AspectRatio(
        aspectRatio: aspectRatio,
        child:
        PageView.builder(
          padEnds: true,
          controller: _pageController,

          itemBuilder: (context, index) {
            return InkWell(
                onTap: ()=> Navigator.push(context, MaterialPageRoute(builder: (context)=>
                    gallery_zoom_slides(
                        arrayImages: arrayImages,
                        currentIndex: index%arrayImages.length)
                )),
                child: Utils.funcDisplayImage(arrayImages[index%arrayImages.length], boxFit));
          }
      ));
  }

}


lib/src/Style2.dart
import 'dart:async';
import 'package:dots_indicator/dots_indicator.dart';
import 'package:flexi_image_slider/src/Utils.dart';
import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';
import 'package:gallery_zoom_slides/gallery_zoom_slides.dart';

class Style2 extends StatelessWidget{

  final BuildContext context;
  bool autoScroll;
  final double aspectRatio;
  BoxFit boxFit;
  List<String> arrayImages;

  Style2({required this.context, required this.autoScroll, required this.aspectRatio, required this.boxFit,required this.arrayImages});

  Timer? _timer;
  final PageController _pageController = PageController(initialPage: 999,viewportFraction: 1.0);
  ValueNotifier<int> newIndexValue = ValueNotifier(0);

  @override
  Widget build(BuildContext context) {

    if(autoScroll){
      _timer = Timer.periodic(Duration(seconds: 3), (Timer timer) {
        _pageController.nextPage(duration: const Duration(seconds: 1), curve: Curves.easeInOutSine);
      });}

    return Column(
      children: [
        AspectRatio(
            aspectRatio: aspectRatio,
            child: PageView.builder(
                padEnds: true,
                controller: _pageController,
                onPageChanged: (newIndex){
                  // print("${newIndex%arrayImages.length}");
                  newIndexValue.value = newIndex%arrayImages.length;
                  newIndexValue.notifyListeners();
                },
                itemBuilder: (context, index) {
                  return
                    InkWell(
                      onTap: ()=> Navigator.push(context, MaterialPageRoute(builder: (context)=>
                          gallery_zoom_slides(
                              arrayImages: arrayImages,
                              currentIndex: index%arrayImages.length)
                      )),
                      child:
                      Utils.funcDisplayImage(arrayImages[index%arrayImages.length], boxFit),);
                }
            )
        ),
        Padding(padding: EdgeInsets.only(top: 8,bottom: 8),
          child: ValueListenableBuilder(
              valueListenable: newIndexValue,
              builder: (context,value,child){
                  return DotsIndicator(
                    dotsCount: arrayImages.length,
                    position: double.parse(
                        "${value%arrayImages.length}"),
                    decorator: DotsDecorator(
                      size: const Size.square(5.0),
                      activeSize: Size(15.0, 5.0),
                      color: Theme.of(context).colorScheme.onBackground.withOpacity(0.37),
                      // Inactive color
                      activeColor: Theme.of(context).colorScheme.primary,
                      activeShape: RoundedRectangleBorder(
                          borderRadius: BorderRadius.circular(5.0)),
                    ),
                  );
              }),
        )
      ],
    );
  }

}


lib/src/Style3.dart
import 'dart:async';
import 'package:flexi_image_slider/src/Utils.dart';
import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';
import 'package:gallery_zoom_slides/gallery_zoom_slides.dart';

class Style3 extends StatelessWidget{

  final BuildContext context;
  bool autoScroll;
  final double aspectRatio;
  BoxFit boxFit;
  List<String> arrayImages;

  Style3({required this.context, required this.autoScroll, required this.aspectRatio, required this.boxFit,required this.arrayImages});

  Timer? _timer;
  final PageController _pageController = PageController(initialPage: 999,viewportFraction: 0.9);

  @override
  Widget build(BuildContext context) {

    if(autoScroll){
      _timer = Timer.periodic(Duration(seconds: 3), (Timer timer) {
        _pageController.nextPage(duration: const Duration(seconds: 1), curve: Curves.easeInOutSine);
      });}

    return
      AspectRatio(
        aspectRatio: aspectRatio,
        child:
        PageView.builder(
          padEnds: true,
          controller: _pageController,

           itemBuilder: (context, index) {
            return Padding(
              padding: const EdgeInsets.only(left: 5,right: 5),
              child:
             InkWell(
             onTap: ()=> Navigator.push(context, MaterialPageRoute(builder: (context)=>
             gallery_zoom_slides(
             arrayImages: arrayImages,
             currentIndex: index%arrayImages.length)
             )),
             child:
              Utils.funcDisplayImage(arrayImages[index%arrayImages.length], boxFit),),
            );
          }
      ));
  }

}


lib/src/flexi_image_slider_base.dart
import 'dart:async';
import 'package:dots_indicator/dots_indicator.dart';
import 'package:flutter/material.dart';
import 'Utils.dart';

enum IndicatorPosition { none, overImage, afterImage }
enum IndicatorAlignment { left, right, center }

class flexi_image_slider extends StatelessWidget{

  BuildContext context;
  final bool autoScroll;
  double aspectRatio;
  final BoxFit boxFit;
  List<String> arrayImages;
  final double viewportFraction;
  final Function? onTap;
  final IndicatorPosition indicatorPosition;
  final IndicatorAlignment indicatorAlignment;
  final Duration duration;
  final double borderRadius;
  final Color indicatorActiveColor;
  final Color indicatorDeactiveColor;

   flexi_image_slider({
     super.key,
     required this.context,
     this.autoScroll = true,
     required this.aspectRatio,
     this.boxFit = BoxFit.fitWidth,
     required this.arrayImages,
     this.viewportFraction = 1.0,
     this.onTap,
     this.indicatorPosition = IndicatorPosition.afterImage,
     this.indicatorAlignment = IndicatorAlignment.center,
     this.duration = const Duration(seconds: 3),
     this.borderRadius = 0.0,
     this.indicatorActiveColor = Colors.black,
     this.indicatorDeactiveColor = Colors.grey,
   });

  Timer? timer;
  PageController pageController = PageController();
  ValueNotifier<int> newIndexValue = ValueNotifier(0);

  Widget funcDotsIndicator(){
    return Padding(padding: EdgeInsets.only(top: 8,bottom: 8),
      child: ValueListenableBuilder(
          valueListenable: newIndexValue,
          builder: (context,value,child){
            return Align(
              alignment: indicatorAlignment == IndicatorAlignment.left ? Alignment.centerLeft : indicatorAlignment == IndicatorAlignment.right ? Alignment.centerRight : Alignment.center,
              child: DotsIndicator(
                dotsCount: arrayImages.length,
                position: double.parse(
                    "${value%arrayImages.length}"),
                decorator: DotsDecorator(
                  size: const Size.square(5.0),
                  activeSize: Size(15.0, 5.0),
                  color: indicatorDeactiveColor,
                  activeColor: indicatorActiveColor,
                  activeShape: RoundedRectangleBorder(
                      borderRadius: BorderRadius.circular(5.0)),
                ),
              ),
            );
          }),
    );
  }

  @override
  Widget build(BuildContext context) {

   pageController = PageController(initialPage: 999,viewportFraction: viewportFraction!);

   if(autoScroll){
     timer = Timer.periodic(duration, (Timer timer) {
       pageController.nextPage(duration: const Duration(seconds: 1), curve: Curves.easeInOutSine);
     });}

    return

      Column(
        children: [

          AspectRatio(
              aspectRatio: aspectRatio,
              child:Stack(
                children: [

                  Positioned.fill(
                      child: PageView.builder(
                          padEnds: true,
                          controller: pageController,
                          onPageChanged: (newIndex){
                            newIndexValue.value = newIndex;
                            newIndexValue.notifyListeners();
                          },
                          itemBuilder: (context, index) {
                            return InkWell(
                                onTap: ()=>onTap!(index%arrayImages.length),
                                child: Padding(
                                  padding: viewportFraction == 1.0 ? EdgeInsets.zero : const EdgeInsets.only(left: 5,right: 5),
                                  child: ClipRRect(
                                      borderRadius: BorderRadius.circular(borderRadius),
                                      child: Utils.funcDisplayImage(arrayImages[index%arrayImages.length], boxFit!)),
                                ));
                          }
                      )
                  ),

                  Positioned(
                    left: 0,right: 0,bottom: 0,
                    child:  indicatorPosition ==  IndicatorPosition.overImage ?
                    funcDotsIndicator()
                        :
                    const SizedBox(),
                  )

                ],
              )
          ),

          indicatorPosition ==  IndicatorPosition.afterImage ?
          funcDotsIndicator()
              :
          const SizedBox()

        ],
      );
  }
}

example/lib/main.dart
import 'package:cached_network_image/cached_network_image.dart';
import 'package:flutter/material.dart';
import 'package:flexi_image_slider/flexi_image_slider.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        // This is the theme of your application.
        //
        // Try running your application with "flutter run". You'll see the
        // application has a blue toolbar. Then, without quitting the app, try
        // changing the primarySwatch below to Colors.green and then invoke
        // "hot reload" (press "r" in the console where you ran "flutter run",
        // or simply save your changes to "hot reload" in a Flutter IDE).
        // Notice that the counter didn't reset back to zero; the application
        // is not restarted.
        primarySwatch: Colors.blue,
      ),
      home: const MyHomePage(title: 'Flexi Image Slider'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({super.key, required this.title});

  // This widget is the home page of your application. It is stateful, meaning
  // that it has a State object (defined below) that contains fields that affect
  // how it looks.

  // This class is the configuration for the state. It holds the values (in this
  // case the title) provided by the parent (in this case the App widget) and
  // used by the build method of the State. Fields in a Widget subclass are
  // always marked "final".

  final String title;

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {

  List<String> arrayImages = [

    "https://rukminim2.flixcart.com/image/832/832/xif0q/t-shirt/5/9/6/m-all-rbc-white-one-nb-nicky-boy-original-imagpycw749kqzdz.jpeg?q=70",
    "https://rukminim2.flixcart.com/image/832/832/xif0q/t-shirt/w/8/l/l-all-rbc-white-one-nb-nicky-boy-original-imagpyct8f5hhauz.jpeg?q=70",
    "https://rukminim2.flixcart.com/image/832/832/xif0q/t-shirt/4/2/c/s-all-rbc-one-nb-nicky-boy-original-imagpycxgmnxretx.jpeg?q=70",
    "https://rukminim2.flixcart.com/image/832/832/xif0q/t-shirt/8/3/7/l-all-rbc-white-one-nb-nicky-boy-original-imagpycbphvjfmk4.jpeg?q=70"

  ];

  @override
  Widget build(BuildContext context) {

    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: Padding(
        padding: const EdgeInsets.only(left: 10,right: 10),
        child: flexi_image_slider(

          //required fields ---------

          context: context,
          aspectRatio: 16/14,
          arrayImages: arrayImages,

          //optional fields ------------

          autoScroll: true,
          viewportFraction: 1.0,
          boxFit: BoxFit.fill,
          indicatorPosition: IndicatorPosition.overImage,//IndicatorPosition.afterImage,IndicatorPosition.overImage,IndicatorPosition.none
          indicatorAlignment: IndicatorAlignment.left,//IndicatorAlignment.left,IndicatorAlignment.right,IndicatorAlignment.center
          duration: const Duration(seconds: 4),
          indicatorActiveColor: Colors.blue,
          indicatorDeactiveColor: Colors.grey,
          borderRadius: 0,
          onTap: (int index){
            print("$index index clicked");
            //handle your click events
          },
        ),
      ),
    );
  }


}
